csrc/disjoint_set.h

```diff
@@ -210,64 +287,78 @@ class DisjointSets {
   }
 
   // Initializes a new set for provided entry
-  //
-  // TODO: Return iterator
-  void initializeSet(T entry) {
-    if (disjoint_set_maps_.find(entry) != disjoint_set_maps_.end()) {
-      return;
+  std::pair<
+      typename std::unordered_map<
+          T,
+          std::shared_ptr<VectorOfUniqueEntries<T, Hash>>,
+          Hash>::iterator,
+      bool>
+  initializeSet(T entry) {
+    auto disjoint_set_maps_it = disjoint_set_maps_.find(entry);
+    if (disjoint_set_maps_it != disjoint_set_maps_.end()) {
+      return std::make_pair(disjoint_set_maps_it, false);
     }
 
     disjoint_sets_.push_back(
         std::make_shared<VectorOfUniqueEntries<T, Hash>>());
     disjoint_sets_.back()->pushBack(entry);
-    disjoint_set_maps_.emplace(std::make_pair(entry, disjoint_sets_.back()));
+    return disjoint_set_maps_.emplace(
+        std::make_pair(entry, disjoint_sets_.back()));
   }
 
   // Adds all of the disjoint set belonging to entry1 to the disjoint set
   // belonging to entry0, maps all entries of disjoint set belonging to entry1
   // to entry0, removes original disjoint set belonging to entry1.
   void mapEntries(T entry0, T entry1) {
+    if (entry0 == entry1) {
```

Comment:
I'm curious to talk more about this. Run to run determinism is why we maintain a vector. It's very important for debugging and consistency that the approach is run to run deterministic. Curious what a union-find structure is though.

Reply:
I like the idea but I think it might be too early to try and do this approach. Generating the IterDomains is pretty challenging and is an iterative process where a concept like iterator can't be completely invalidated when we perform a merge. That's one reason in disjoint_set I'm not reusing any of the groups, but I remove the old groups and put in a brand new group. That's because those old groups (from what I can tell) need to be valid after that merge for an unspecified amount of time (i.e. why they're shared memory pointers).

The process of how the graph is built motivates how the data structure works. With that said, I think once we have an implementation that's fully functional we can revisit this and work through the exact invalidation scheme we can tolerate for a different approach. So I think this response isn't more of an if we improve disjoint_set, but when ;-)

p.s. this response assumes that the approach would invalidate all iterators (references to groups) when a merge is performed.